core/job/service/job_service.go

```diff
@@ -331,20 +334,42 @@ func (j JobService) Validate(ctx context.Context, jobTenant tenant.Tenant, jobSp
 	if err != nil {
 		return err
 	}
-	jobMap := make(map[job.Name]*job.Job)
-	destinationToJobsMap := make(map[job.ResourceURN][]*job.Job)
-	for _, jobEntity := range jobsInProject {
-		jobMap[jobEntity.Spec().Name()] = jobEntity
-		if _, ok := destinationToJobsMap[jobEntity.Destination()]; !ok {
-			destinationToJobsMap[jobEntity.Destination()] = []*job.Job{}
+	jobsInProjectWithUpstreams, err := job.Jobs(jobsInProject).GetJobsWithUnresolvedUpstreams()
```

Comment:
if possible we can extract out:
- fetching `jobsInProject`
- getting `jobsInProjectWithUpstreams`
- maps preparation
into a single function to make `Validate` easier to read.